File: src/model/map/MapConfiguration.scala
package model.map

import java.io._
import java.nio.file.Paths
import java.util.Scanner

import scala.collection.immutable.HashMap

object MapConfiguration {
  // Turns a text based configuration file into a MapConfiguration object as well as serializing
  def createMapConfigFromFile(in: InputStream, outfileName: String): MapConfiguration = {
    val reader = new Scanner(in)
    val rows: IndexedSeq[List[(String, NodeType)]] = for (suffix <- GameMap.nodeSuffixes) yield {
      val line: String = reader.nextLine()
      val lineAsChars: List[Char] = line.toList
      val typeIdentifierAndPrefix: List[(Char, Char)] = lineAsChars.zip(GameMap.nodePrefixes)
      val rowAsMappedValues: List[(String, NodeType)] = typeIdentifierAndPrefix map {
        case (typeIdentifier, prefix) => (prefix + suffix, GameMap.characterToNodeType(typeIdentifier))
      }
      rowAsMappedValues
    }
    val mappedElements: List[(String, NodeType)] = rows.flatten.toList
    val configuration = new MapConfiguration(HashMap(mappedElements: _*))
    val outStream = new ObjectOutputStream(new FileOutputStream(outfileName))
    outStream.writeObject(configuration)
    outStream.close()
    configuration
  }


  def readConfigurationFromFile(inputStream: InputStream): MapConfiguration = {
    val input = new ObjectInputStream(inputStream)
    val obj = input.readObject()
    obj match {
      case cfg: MapConfiguration => cfg
      case _ => sys.error("Type mismatch when reading from file")
    }
  }
}

class MapConfiguration(tileConfig: HashMap[String, NodeType]) extends Serializable {
  def apply(s: String): NodeType = tileConfig(s)
}

File: outdated-code/MapConfiguration.java
//package model.map;
//
//import java.io.*;
//import java.nio.file.Paths;
//import java.util.HashMap;
//import java.util.Scanner;
//
//public class MapConfiguration implements Serializable{
//	private static char[] nodePrefixes = {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W'};
//	private static String[] nodeSuffixes = {"01", "02", "03", "04", "05", "06", "07", "08", "09", "10", "11", "12", "13", "14"};
//	private HashMap<String, NodeType> config = new HashMap<>();
//
//	public NodeType getNodeType(String s){
//		return config.get(s);
//	}
//
//
//	public static void createMapConfigFromFile(InputStream in, String out){
//		MapConfiguration newConfig = new MapConfiguration();
//		Scanner reader;
//		int rowPointer = 0;
//		try{
//            reader = new Scanner(in);
//
//            while(reader.hasNextLine()){
//                String currLine = reader.nextLine();
//                for(int i = 0; i < currLine.length(); i++){
//                    String nodeName = nodePrefixes[i] + nodeSuffixes[rowPointer];
//                    newConfig.config.put(nodeName, charCorrespondence.get(currLine.charAt(i)));
//                    System.out.println(nodeName + " is " + newConfig.config.get(nodeName));
//                }
//                rowPointer++;
//            }
//
//            ObjectOutputStream outStream = new ObjectOutputStream(new BufferedOutputStream(new FileOutputStream(new File("EFTAIOS/src/main/resources/" + out))));
//            outStream.writeObject(newConfig);
//            outStream.close();
//		}catch(IOException e){
//			e.printStackTrace();
//			System.exit(1);
//		}
//	}
//	public static MapConfiguration loadFromFile(InputStream s){
//		try{
//			ObjectInputStream inStream = new ObjectInputStream(s);
//			MapConfiguration config = (MapConfiguration)inStream.readObject();
//			return config;
//		}catch(IOException e){
//			e.printStackTrace();
//			System.exit(1);
//		}catch(ClassNotFoundException e){
//			e.printStackTrace();
//			System.exit(1);
//		}
//		//THIS SHOULD NEVER HAPPEN
//		return null;
//	}
//}
